### PR TITLE
Add success and error snackbars to the templates list page

### PR DESCRIPTION
**Pull request by kevin940726**
## Description

Part of https://github.com/WordPress/gutenberg/issues/36597.

Add success and error snackbars to the templates list page. It will now:

1. Show a success notice when a template(-part) is successfully created or removed.
2. Show an error notice when there's something wrong in (1).

Since that the snackbars have similar background as the navigation panel, they will be pushed right a little bit when the panel is opened.

## How has this been tested?

1. Activate `tt1-blocks` theme
2. Go to Appearance -> Editor -> top left corner W menu -> Templates / Template Parts
3. Try creating a template or a template part
4. You won't notice the snackbar if the navigation is very fast, but with a poor network you will notice something like "Template created. Navigating to the editor..."
5. Try removing a template or a template part
6. The snackbar will show "{template title} removed"
7. If something went wrong when retrying (3) and (5) (like adding intentional `throw`), you should notice error snackbars.

## Screenshots 
https://user-images.githubusercontent.com/7753001/143198279-021d1e9f-376f-4d5f-88a8-f4ceb96015c9.mp4

## Types of changes

New feature

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/edit-site/src/components/add-new-template/new-template-part.js
Comment (kevin940726) [inline packages/edit-site/src/components/add-new-template/new-template-part.js]: Can't decide whether we should automatically navigate the user to the created template or show a "View template" link in the success snackbar. @jameskoster WDYT?

File: packages/edit-site/src/store/actions.js
Comment (kevin940726) [inline packages/edit-site/src/store/actions.js:115]: For some reasons, network failure won't throw an error here 🤔 . Maybe we should rewrite this into plain `apiFetch`.

File: packages/edit-site/src/components/add-new-template/new-template-part.js
Comment (Mamaduka) [reply packages/edit-site/src/components/add-new-template/new-template-part.js]: We display success notice after "navigating" users to a template in post editor Template Mode.

Comment (jameskoster) [reply packages/edit-site/src/components/add-new-template/new-template-part.js]: I'm not entirely sure we need a Snackbar on _creation_, but let's try it after sending people to the editor like @Mamaduka suggested before deciding.

**Comment by jameskoster** [comment]
Since we don't have a trash feature, it would be excellent to add an undo affordance to the deletion Snackbar.

We should also display a snackbar when customisations are cleared – again with undo – so it may be worth holding off on merging this until #36802 is closed?

File: packages/edit-site/src/components/list/index.js
Comment (Mamaduka) [inline packages/edit-site/src/components/list/index.js]: Should we move this logic to `InterfaceSkeleton` and I think we can avoid wrapper div.

File: package-lock.json
Comment (Mamaduka) [inline package-lock.json]: See: https://github.com/WordPress/gutenberg/pull/36511#discussion_r753375631

File: packages/edit-site/src/components/list/index.js
Comment (Mamaduka) [reply packages/edit-site/src/components/list/index.js]: I changed this in aba861d, but feel free to revert.

**Comment by noisysocks** [comment]
I'm not a huge fan of the "Template created. Navigating to the editor..." notice. On Firefox the page freezes while that notice is fading in so the whole thing looks kinda strange. I think don't bother with a notice on creation or, as discussed above, show it after the navigation.

I tested _Clear customizations_ on this branch and didn't see a notice. We should add that.

File: packages/edit-site/src/components/add-new-template/new-template-part.js
Comment (noisysocks) [inline packages/edit-site/src/components/add-new-template/new-template-part.js]:
```suggestion
					: __( 'An error occurred while creating the template part.' );
```

File: packages/edit-site/src/components/add-new-template/new-template-part.js
Comment (noisysocks) [inline packages/edit-site/src/components/add-new-template/new-template-part.js:28]: How would this happen? 🙂 I guess it's okay to leave it just in case.

File: packages/edit-site/src/components/add-new-template/new-template.js
Comment (noisysocks) [inline packages/edit-site/src/components/add-new-template/new-template.js]:
```suggestion
					: __( 'An error occurred while creating the template.' );
```

File: packages/edit-site/src/store/actions.js
Comment (noisysocks) [reply packages/edit-site/src/store/actions.js:115]: That sounds like a `core-data` bug?

File: packages/edit-site/src/store/actions.js
Comment (noisysocks) [inline packages/edit-site/src/store/actions.js]:
```suggestion
				: __( 'An error occurred while removing the template.' );
```

**Review by noisysocks — COMMENTED** [review]
Code looks good!

File: packages/edit-site/src/store/actions.js
Comment (jameskoster) [reply packages/edit-site/src/store/actions.js]: Let's use "deleting" rather than "removing". It's more explicit and aligns with https://github.com/WordPress/gutenberg/pull/36815

File: packages/edit-site/src/store/actions.js
Comment (talldan) [reply packages/edit-site/src/store/actions.js:115]: There is a `getLastEntityDeleteError` selector in the core store.

File: packages/edit-site/src/components/add-new-template/new-template-part.js
Comment (kevin940726) [reply packages/edit-site/src/components/add-new-template/new-template-part.js:28]: Yeah, it will never happen for regular users. Maybe we should just remove this check altogether.

File: packages/edit-site/src/store/actions.js
Comment (kevin940726) [reply packages/edit-site/src/store/actions.js:115]: Yeah, seems like a `core-data` bug. Not sure how we can use `getLastEntityDeleteError` here? Better leave it for another PR anyway.

File: packages/edit-site/src/components/add-new-template/new-template-part.js
Comment (kevin940726) [reply packages/edit-site/src/components/add-new-template/new-template-part.js]: Since we do a full page navigation when creating a template, it's currently not possible to delay the snackbar after the navigation. We could use query params or `sessionStorage` for this though but not trivial either. Should better leave it to another PR if necessary.

File: packages/edit-site/src/store/actions.js
Comment (talldan) [reply packages/edit-site/src/store/actions.js:115]: You'd call the selector after the delete to check if there was an error. I think it takes the same args as `deleteEntityRecord` to make sure the error is related to the particular request to delete something.

File: packages/edit-site/src/index.js
Comment (talldan) [inline packages/edit-site/src/index.js:92]: There's a `useViewportMatch` hook in the compose package, I don't think we need to bring in the viewport dependency directly.

Comment (kevin940726) [reply packages/edit-site/src/index.js:92]: I want to synchronously update the store's default value before rendering the React tree. So that we don't have to put it inside an `useEffect` to schedule it after the first render, causing unnecessary re-render and jump of UI.

Comment (talldan) [reply packages/edit-site/src/index.js:92]: Ok, I see. Might be worth adding a comment here to explain that so that someone like me doesn't refactor it 😄

File: packages/edit-site/src/store/actions.js
Comment (kevin940726) [reply packages/edit-site/src/store/actions.js:115]: I see. Good suggestion!

File: packages/edit-site/src/index.js
Comment (kevin940726) [reply packages/edit-site/src/index.js:92]: Sure, good idea.

File: packages/edit-site/src/store/test/actions.js
Comment (kevin940726) [inline packages/edit-site/src/store/test/actions.js:79]: I removed the unit test because I don't think it's serving any purpose besides slowing us down. We could add it back if the action becomes more complicated. Or if anyone thinks otherwise.

File: packages/edit-site/src/components/add-new-template/new-template-part.js
Comment (talldan) [reply packages/edit-site/src/components/add-new-template/new-template-part.js]: I think we should just remove this notice if there's going to be a navigation.

For a lot of screenreader users this they won't ever hear this message because of the navigation. Visually it's probably hard to spot too.

Comment (kevin940726) [reply packages/edit-site/src/components/add-new-template/new-template-part.js]: Yep, the creation notices should be removed.

I wonder if it goes another way if we're developing a SPA though. Then I think it would make sense to add this notice back? (But probably should also delay it after navigation?)

Comment (talldan) [reply packages/edit-site/src/components/add-new-template/new-template-part.js]: If we want to keep it, I think we should change the exclamation for a period.

I feel like I'm more used to seeing exclamations in error messages to get a user's attention.

Comment (talldan) [reply packages/edit-site/src/components/add-new-template/new-template-part.js]: Ok, I see it has been removed now, I was on a slightly outdated verison of the branch. 👍

**Review by talldan — APPROVED** [review]
LGTM - thanks @kevin940726

File: packages/edit-site/src/store/test/actions.js
Comment (talldan) [reply packages/edit-site/src/store/test/actions.js:79]: Yeah, agree with removing it. There are already tests for deleteEntityRecord, and the notices should ideally be tested with e2e tests if they're to be tested.

**Comment by jameskoster** [comment]
One little inconsistency that'd be nice to fix here is that when you delete a template, we include the template name in the Snackbar, but when you revert a template we do not. I suppose we can address that when we add the undo affordance.

**Pull request merged**
